### PR TITLE
Implement #762

### PR DESCRIPTION
**Pull request by zero-88**
@cushon  #762

### CODE REVIEW

File: core/src/main/java/com/google/errorprone/bugpatterns/LinkedListChecker.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+@AutoService(BugChecker.class)
+@BugPattern(name = "LinkedListChecker", summary = "java.util.LinkedList accessed through via index",
+    severity = ERROR, category = Category.JDK)
+public final class LinkedListChecker extends BugChecker
+    implements MethodInvocationTreeMatcher, InstanceOfTreeMatcher {
+
+  private static final String LINKED_LIST_FQN = "java.util.LinkedList";
+  private static final Matcher<MethodInvocationTree> INSTANCE_LINKEDLIST_MATCHER = Matchers.anyOf(
+      MethodMatchers.instanceMethod().onDescendantOf(LINKED_LIST_FQN).named("add")
```
Comment (ronshapiro) [inline core/src/main/java/com/google/errorprone/bugpatterns/LinkedListChecker.java]: This isn't necessarily bad. `myLinkedList.add(0, item)` is actually probably a good intended use of a `LinkedList`.

Maybe at the least special-case `0`, or even any small contant (or perhaps all constants)?

The same applies for the other methods below.

File: core/src/main/java/com/google/errorprone/bugpatterns/LinkedListChecker.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!INSTANCE_LINKEDLIST_MATCHER.matches(tree, state)) {
+      return NO_MATCH;
+    }
+    return this.describeMatch(tree);
+  }
+
+  @Override
+  public Description matchInstanceOf(InstanceOfTree tree, VisitorState state) {
```
Comment (ronshapiro) [inline core/src/main/java/com/google/errorprone/bugpatterns/LinkedListChecker.java]: Did you mean to remove this (and the interface that requires it)?

**Comment by ronshapiro** [comment]
This check requires people to store references as a `LinkedList<T>` instead of just a plain `List<T>` which we tend to prefer (Effective Java 2nd edition Item 52). In general, the performance of `LinkedList` isn't so great compared to `LinkedList`, so I'm not sure if this will get many hits.

**Comment by zero-88** [comment]
@ronshapiro 
I already updated code. Could you help me to review it again?
Anything misunderstood, I left comment in #762

**Comment by cushon** [comment]
> In general, the performance of `LinkedList` isn't so great compared to [`ArrayList`], so I'm not sure if this will get many hits.

Given that, I think we're better off discouraging uses of `LinkedList` altogether (see [`JdkObsolete`](https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/JdkObsolete.java)), rather than warning about specific uses of `LinkedList`.

**Comment by zero-88** [comment]
Hi @cushon, I agree with your solution. 
However, I just wondering:
- Should we point out exact issue in description? In this case, it is a list method that I checked.
- Seems `JdkObsolete` not yet checked inheritance. E.g: Clazz extends HashTable/LinkedList

**Comment by cushon** [comment]
> Should we point out exact issue in description? In this case, it is a list method that I checked.

I'm not sure if I understand. Are you suggesting having `JdkObsolete` emit more specific diagnostics, or also handling method calls on the obsolete types (like this check does), or something else? My thinking was that checking instantiation (and now extending/implementing) the obsolete classes is enough. We want to prevent new uses of these types from being introduced, but if you're calling `ArrayList#get` it might be because some API you don't have control over returned that type, so it would be annoying to get a warning about it.

> Seems JdkObsolete not yet checked inheritance. E.g: Clazz extends HashTable/LinkedList

Good idea: 319c50539eeab6442963e23a26f8b7119539a778.

**Comment by zero-88** [comment]
Actually, I would like to indicate exact operation that slow down performance than restrict use any class.
For example, linkedlist is faster than arraylist in case of using operation without index. It is good point to use it.
Also, it helps to scan old system when using class operation instead of checking new instance. Somehow, client don't want a change their structure but can refactor some method.
However, the current limitation is we cannot check if var's declared as List interface.
That's my idea.

**Comment by Stephan202** [comment]
>  For example, linkedlist is faster than arraylist in case of using operation without index. It is good point to use it.

Can you expand on that? ([Potentially relevant video](https://www.youtube.com/watch?v=YQs6IC-vgmo))

**Comment by cushon** [comment]
Our benchmarking shows that `ArrayList` or `ArrayDeque` are effectively always a better choice. I don't think it's worth trying to discourage particular uses of `LinkedList` instead of discouraging *all* uses. I can imagine people seeing this check and getting the idea that `LinkedList` is a good choice as long as they don't need to index into it, but it really isn't.

If you'd still like to use this check on your code base, you could add it as a [plugin check](http://errorprone.info/docs/plugins).

**Comment by cushon** [comment]
Also, https://twitter.com/joshbloch/status/583813919019573248.

**Comment by zero-88** [comment]
Thank @cushon and @Stephan202 for interesting information.
I will have a look on plugin check because based on my requirements. 
We can close topic at here 👍

**Pull request closed without merging**